Networking/Networking/ApplicationPassword/RequestAuthenticator.swift

```diff
@@ -0,0 +1,91 @@
+            } else if case let .wporg(username, password, siteAddress) = credentials {
+                return try? DefaultApplicationPasswordUseCase(username: username,
+                                                              password: password,
+                                                              siteAddress: siteAddress)
+            } else {
+                return nil
+            }
+        }()
+        self.applicationPasswordUseCase = useCase
+    }
+
+    func authenticate(_ urlRequest: URLRequest) throws -> URLRequest {
+        guard case let .wporg(_, _, siteAddress) = credentials,
+              let url = urlRequest.url,
+              url.absoluteString.hasPrefix(siteAddress.trimSlashes() + "/" + RESTRequest.Settings.basePath) else {
+            // Handle non-REST requests as before
+            return try authenticateUsingWPCOMTokenIfPossible(urlRequest)
+        }
+
+        return try authenticateUsingApplicationPasswordIfPossible(urlRequest)
+    }
+
+    func generateApplicationPassword() async throws {
+        guard let applicationPasswordUseCase = applicationPasswordUseCase else {
+            throw RequestAuthenticatorError.applicationPasswordUseCaseNotAvailable
+        }
+        let _ = try await applicationPasswordUseCase.generateNewPassword()
+        return
+    }
+}
+
+private extension RequestAuthenticator {
+    /// Attempts creating a request with WPCOM token if possible.
+    ///
+    func authenticateUsingWPCOMTokenIfPossible(_ urlRequest: URLRequest) throws -> URLRequest {
+        if let credentials, case .wpcom = credentials {
+            return try AuthenticatedRequest(credentials: credentials, request: urlRequest).asURLRequest()
+        }
+        return UnauthenticatedRequest(request: urlRequest).asURLRequest()
+    }
+
+    /// Attempts creating a request with application password if possible.
+    ///
+    func authenticateUsingApplicationPasswordIfPossible(_ urlRequest: URLRequest) throws -> URLRequest {
+        guard let applicationPassword = applicationPasswordUseCase?.applicationPassword else {
+            throw RequestAuthenticatorError.applicationPasswordNotAvailable
+        }
+
+        var request = urlRequest
+        request.setValue("application/json", forHTTPHeaderField: "Accept")
+        request.setValue(UserAgent.defaultUserAgent, forHTTPHeaderField: "User-Agent")
+
+        let username = applicationPassword.wpOrgUsername
+        let password = applicationPassword.password.secretValue
+        let loginString = "\(username):\(password)"
+        guard let loginData = loginString.data(using: .utf8) else {
+            return request
+        }
+        let base64LoginString = loginData.base64EncodedString()
+        request.setValue("Basic \(base64LoginString)", forHTTPHeaderField: "Authorization")
+
+        // Cookies from `CookieNonceAuthenticator` should be skipped
+        request.httpShouldHandleCookies = false
```

Comment:
does it make sense to move these auth logic for a REST request to `RESTRequest.asURLRequest`, so that it's similar to `authenticateUsingWPCOMTokenIfPossible` and can be tested separately?

Reply:
Good point. Done in 5965160b490ecd85069e2d39ab9f9c88f6cb96fa